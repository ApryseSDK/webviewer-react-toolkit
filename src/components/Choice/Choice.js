'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.Choice = void 0;
var tslib_1 = require('tslib');
var classnames_1 = tslib_1.__importDefault(require('classnames'));
var react_1 = tslib_1.__importStar(require('react'));
var hooks_1 = require('../../hooks');
var useID_1 = require('../../hooks/useID');
var Icon_1 = require('../Icon');
exports.Choice = (0, react_1.forwardRef)(function(_a, ref) {
  var label = _a.label,
    leftLabel = _a.leftLabel,
    className = _a.className,
    children = _a.children,
    id = _a.id,
    radio = _a.radio,
    isSwitch = _a.isSwitch,
    center = _a.center,
    disabledLabelChange = _a.disabledLabelChange,
    onChange = _a.onChange,
    onFocus = _a.onFocus,
    onBlur = _a.onBlur,
    props = tslib_1.__rest(_a, [
      'label',
      'leftLabel',
      'className',
      'children',
      'id',
      'radio',
      'isSwitch',
      'center',
      'disabledLabelChange',
      'onChange',
      'onFocus',
      'onBlur',
    ]);
  var inputRef = (0, react_1.useRef)(null);
  (0, react_1.useImperativeHandle)(ref, function() {
    return inputRef.current;
  });
  var isUserTabbing = (0, hooks_1.useAccessibleFocus)();
  var _b = (0, hooks_1.useFocus)(onFocus, onBlur),
    focused = _b.focused,
    handleOnFocus = _b.handleOnFocus,
    handleOnBlur = _b.handleOnBlur;
  var choiceID = (0, useID_1.useID)(id);
  var _c = (0, react_1.useState)(function() {
      var _a, _b, _c;
      return (_c =
        (_a = props.checked) !== null && _a !== void 0
          ? _a
          : (_b = inputRef.current) === null || _b === void 0
          ? void 0
          : _b.checked) !== null && _c !== void 0
        ? _c
        : false;
    }),
    checked = _c[0],
    setChecked = _c[1];
  (0, react_1.useEffect)(
    function() {
      if (props.checked !== undefined) setChecked(props.checked);
    },
    [props.checked],
  );
  var handleOnChange = function(event) {
    if (props.checked === undefined) setChecked(event.target.checked);
    onChange === null || onChange === void 0 ? void 0 : onChange(event);
  };
  // HACK: since there is no way to detect that a radio button is being
  // unchecked due to form activity, we subscribe all radio buttons to an
  // observable. When checked changes and props.checked is undefined, this
  // will trigger the observable which will ping all other subscribers to
  // check if they have become unchecked, and if so change their internal
  // state.
  (0, react_1.useEffect)(
    function() {
      if (props.name && radio) {
        return observable.subscribe(props.name, function() {
          if (inputRef.current && inputRef.current.checked !== checked) {
            setChecked(inputRef.current.checked);
          }
        });
      }
      return;
    },
    [checked, props.name, radio],
  );
  var choiceClass = (0, classnames_1.default)(
    'ui__base ui__choice',
    {
      'ui__choice--radio': radio,
      'ui__choice--leftLabel': leftLabel,
      'ui__choice--checked': checked,
      'ui__choice--center': center,
      'ui__choice--disabled': props.disabled,
    },
    className,
  );
  var inputClass = (0, classnames_1.default)('ui__choice__input', {
    'ui__choice__input--switch': isSwitch,
  });
  var checkClass = isSwitch
    ? (0, classnames_1.default)('ui__choice__input__switch', {
        'ui__choice__input__switch--checked': checked,
        'ui__choice__input__switch--disabled': props.disabled,
        'ui__choice__input__switch--focus': isUserTabbing && focused,
      })
    : (0, classnames_1.default)('ui__choice__input__check', {
        'ui__choice__input__check--checked': checked,
        'ui__choice__input__check--disabled': props.disabled,
        'ui__choice__input__check--focus': isUserTabbing && focused,
      });
  var labelClass = (0, classnames_1.default)('ui__choice__label', {
    'ui__choice__label--disabled': props.disabled && disabledLabelChange,
  });
  var labelElement = (0, react_1.useMemo)(
    function() {
      if (!label) return undefined;
      return react_1.default.createElement(
        'label',
        { className: labelClass, htmlFor: choiceID },
        label,
      );
    },
    [choiceID, label, labelClass],
  );
  return react_1.default.createElement(
    'span',
    { className: choiceClass },
    leftLabel ? labelElement : undefined,
    react_1.default.createElement(
      'span',
      { className: inputClass },
      isSwitch
        ? react_1.default.createElement(
            'div',
            { className: checkClass },
            react_1.default.createElement('div', {
              className: 'ui__choice__input__toggle',
            }),
          )
        : react_1.default.createElement(
            'div',
            { className: checkClass },
            checked && !radio
              ? react_1.default.createElement(Icon_1.Icon, {
                  icon: 'Check',
                  className: 'ui__choice__input__icon',
                })
              : undefined,
          ),
      react_1.default.createElement(
        'input',
        tslib_1.__assign({}, props, {
          id: choiceID,
          type: radio ? 'radio' : 'checkbox',
          onChange: handleOnChange,
          ref: inputRef,
          onFocus: handleOnFocus,
          onBlur: handleOnBlur,
        }),
        children,
      ),
    ),
    !leftLabel ? labelElement : undefined,
  );
});
/**
 * Observable helper to notify radio buttons that they have become unchecked.
 */
var RadioObservable = /** @class */ (function() {
  function RadioObservable() {
    this._subscribers = [];
  }
  RadioObservable.prototype.subscribe = function(name, subscriber) {
    this._trigger(name);
    this._subscribers.push({ name: name, subscriber: subscriber });
    return this._unsubscribe(subscriber);
  };
  RadioObservable.prototype._trigger = function(name) {
    this._subscribers.forEach(function(s) {
      if (name === s.name) s.subscriber();
    });
  };
  RadioObservable.prototype._unsubscribe = function(subscriber) {
    var _this = this;
    return function() {
      _this._subscribers = _this._subscribers.filter(function(s) {
        return s.subscriber !== subscriber;
      });
    };
  };
  return RadioObservable;
})();
var observable = new RadioObservable();
//# sourceMappingURL=Choice.js.map
