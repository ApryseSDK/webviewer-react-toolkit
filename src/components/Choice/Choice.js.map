{"version":3,"file":"Choice.js","sourceRoot":"","sources":["Choice.tsx"],"names":[],"mappings":";;;;AAAA,kEAAoC;AACpC,qDAUe;AACf,qCAA2D;AAC3D,2CAA0C;AAE1C,gCAA+B;AAgClB,QAAA,MAAM,GAAG,IAAA,kBAAU,EAC9B,UACE,EAcC,EACD,GAAG;IAdD,IAAA,KAAK,WAAA,EACL,SAAS,eAAA,EACT,SAAS,eAAA,EACT,QAAQ,cAAA,EACR,EAAE,QAAA,EACF,KAAK,WAAA,EACL,QAAQ,cAAA,EACR,MAAM,YAAA,EACN,mBAAmB,yBAAA,EACnB,QAAQ,cAAA,EACR,OAAO,aAAA,EACP,MAAM,YAAA,EACH,KAAK,sBAbV,4IAcC,CADS;IAIV,IAAM,QAAQ,GAAG,IAAA,cAAM,EAAmB,IAAI,CAAC,CAAC;IAChD,IAAA,2BAAmB,EAAC,GAAG,EAAE,cAAM,OAAA,QAAQ,CAAC,OAA2B,EAApC,CAAoC,CAAC,CAAC;IAErE,IAAM,aAAa,GAAG,IAAA,0BAAkB,GAAE,CAAC;IACrC,IAAA,KAA2C,IAAA,gBAAQ,EAAC,OAAO,EAAE,MAAM,CAAC,EAAlE,OAAO,aAAA,EAAE,aAAa,mBAAA,EAAE,YAAY,kBAA8B,CAAC;IAE3E,IAAM,QAAQ,GAAG,IAAA,aAAK,EAAC,EAAE,CAAC,CAAC;IAErB,IAAA,KAAwB,IAAA,gBAAQ,EAAU,8BAAM,OAAA,MAAA,MAAA,KAAK,CAAC,OAAO,mCAAI,MAAA,QAAQ,CAAC,OAAO,0CAAE,OAAO,mCAAI,KAAK,CAAA,EAAA,CAAC,EAAnG,OAAO,QAAA,EAAE,UAAU,QAAgF,CAAC;IAC3G,IAAA,iBAAS,EAAC;QACR,IAAI,KAAK,CAAC,OAAO,KAAK,SAAS;YAAE,UAAU,CAAC,KAAK,CAAC,OAAO,CAAC,CAAC;IAC7D,CAAC,EAAE,CAAC,KAAK,CAAC,OAAO,CAAC,CAAC,CAAC;IAEpB,IAAM,cAAc,GAAG,UAAC,KAAoC;QAC1D,IAAI,KAAK,CAAC,OAAO,KAAK,SAAS;YAAE,UAAU,CAAC,KAAK,CAAC,MAAM,CAAC,OAAO,CAAC,CAAC;QAClE,QAAQ,aAAR,QAAQ,uBAAR,QAAQ,CAAG,KAAK,CAAC,CAAC;IACpB,CAAC,CAAC;IAEF,qEAAqE;IACrE,uEAAuE;IACvE,wEAAwE;IACxE,uEAAuE;IACvE,uEAAuE;IACvE,SAAS;IACT,IAAA,iBAAS,EAAC;QACR,IAAI,KAAK,CAAC,IAAI,IAAI,KAAK,EAAE;YACvB,OAAO,UAAU,CAAC,SAAS,CAAC,KAAK,CAAC,IAAI,EAAE;gBACtC,IAAI,QAAQ,CAAC,OAAO,IAAI,QAAQ,CAAC,OAAO,CAAC,OAAO,KAAK,OAAO,EAAE;oBAC5D,UAAU,CAAC,QAAQ,CAAC,OAAO,CAAC,OAAO,CAAC,CAAC;iBACtC;YACH,CAAC,CAAC,CAAC;SACJ;QACD,OAAO;IACT,CAAC,EAAE,CAAC,OAAO,EAAE,KAAK,CAAC,IAAI,EAAE,KAAK,CAAC,CAAC,CAAC;IAEjC,IAAM,WAAW,GAAG,IAAA,oBAAU,EAC5B,qBAAqB,EACrB;QACE,mBAAmB,EAAE,KAAK;QAC1B,uBAAuB,EAAE,SAAS;QAClC,qBAAqB,EAAE,OAAO;QAC9B,oBAAoB,EAAE,MAAM;QAC5B,sBAAsB,EAAE,KAAK,CAAC,QAAQ;KACvC,EACD,SAAS,CACV,CAAC;IAEF,IAAM,UAAU,GAAG,IAAA,oBAAU,EAAC,mBAAmB,EAAE;QACjD,2BAA2B,EAAE,QAAQ;KACtC,CAAC,CAAC;IAEH,IAAM,UAAU,GAAG,QAAQ;QACzB,CAAC,CAAC,IAAA,oBAAU,EAAC,2BAA2B,EAAE;YACtC,oCAAoC,EAAE,OAAO;YAC7C,qCAAqC,EAAE,KAAK,CAAC,QAAQ;YACrD,kCAAkC,EAAE,aAAa,IAAI,OAAO;SAC7D,CAAC;QACJ,CAAC,CAAC,IAAA,oBAAU,EAAC,0BAA0B,EAAE;YACrC,mCAAmC,EAAE,OAAO;YAC5C,oCAAoC,EAAE,KAAK,CAAC,QAAQ;YACpD,iCAAiC,EAAE,aAAa,IAAI,OAAO;SAC5D,CAAC,CAAC;IAEP,IAAM,UAAU,GAAG,IAAA,oBAAU,EAAC,mBAAmB,EAAE;QACjD,6BAA6B,EAAE,KAAK,CAAC,QAAQ,IAAI,mBAAmB;KACrE,CAAC,CAAC;IAEH,IAAM,YAAY,GAAG,IAAA,eAAO,EAAC;QAC3B,IAAI,CAAC,KAAK;YAAE,OAAO,SAAS,CAAC;QAC7B,OAAO,CACL,yCAAO,SAAS,EAAE,UAAU,EAAE,OAAO,EAAE,QAAQ,IAC5C,KAAK,CACA,CACT,CAAC;IACJ,CAAC,EAAE,CAAC,QAAQ,EAAE,KAAK,EAAE,UAAU,CAAC,CAAC,CAAC;IAElC,OAAO,CACL,wCAAM,SAAS,EAAE,WAAW;QACzB,SAAS,CAAC,CAAC,CAAC,YAAY,CAAC,CAAC,CAAC,SAAS;QACrC,wCAAM,SAAS,EAAE,UAAU;YACxB,QAAQ,CAAC,CAAC,CAAC,CACV,uCAAK,SAAS,EAAE,UAAU;gBACxB,uCAAK,SAAS,EAAC,2BAA2B,GAAG,CACzC,CACP,CAAC,CAAC,CAAC,CACF,uCAAK,SAAS,EAAE,UAAU,IACvB,OAAO,IAAI,CAAC,KAAK,CAAC,CAAC,CAAC,8BAAC,WAAI,IAAC,IAAI,EAAC,OAAO,EAAC,SAAS,EAAC,yBAAyB,GAAG,CAAC,CAAC,CAAC,SAAS,CACtF,CACP;YACD,4DACM,KAAK,IACT,EAAE,EAAE,QAAQ,EACZ,IAAI,EAAE,KAAK,CAAC,CAAC,CAAC,OAAO,CAAC,CAAC,CAAC,UAAU,EAClC,QAAQ,EAAE,cAAc,EACxB,GAAG,EAAE,QAAQ,EACb,OAAO,EAAE,aAAa,EACtB,MAAM,EAAE,YAAY,KAEnB,QAAQ,CACH,CACH;QACN,CAAC,SAAS,CAAC,CAAC,CAAC,YAAY,CAAC,CAAC,CAAC,SAAS,CACjC,CACR,CAAC;AACJ,CAAC,CACF,CAAC;AAEF;;GAEG;AACH;IAGE;QACE,IAAI,CAAC,YAAY,GAAG,EAAE,CAAC;IACzB,CAAC;IAED,mCAAS,GAAT,UAAU,IAAY,EAAE,UAAsB;QAC5C,IAAI,CAAC,QAAQ,CAAC,IAAI,CAAC,CAAC;QAEpB,IAAI,CAAC,YAAY,CAAC,IAAI,CAAC,EAAE,IAAI,MAAA,EAAE,UAAU,YAAA,EAAE,CAAC,CAAC;QAC7C,OAAO,IAAI,CAAC,YAAY,CAAC,UAAU,CAAC,CAAC;IACvC,CAAC;IAEO,kCAAQ,GAAhB,UAAiB,IAAY;QAC3B,IAAI,CAAC,YAAY,CAAC,OAAO,CAAC,UAAC,CAAC;YAC1B,IAAI,IAAI,KAAK,CAAC,CAAC,IAAI;gBAAE,CAAC,CAAC,UAAU,EAAE,CAAC;QACtC,CAAC,CAAC,CAAC;IACL,CAAC;IAEO,sCAAY,GAApB,UAAqB,UAAsB;QAA3C,iBAIC;QAHC,OAAO;YACL,KAAI,CAAC,YAAY,GAAG,KAAI,CAAC,YAAY,CAAC,MAAM,CAAC,UAAC,CAAC,IAAK,OAAA,CAAC,CAAC,UAAU,KAAK,UAAU,EAA3B,CAA2B,CAAC,CAAC;QACnF,CAAC,CAAC;IACJ,CAAC;IACH,sBAAC;AAAD,CAAC,AAzBD,IAyBC;AAED,IAAM,UAAU,GAAG,IAAI,eAAe,EAAE,CAAC"}