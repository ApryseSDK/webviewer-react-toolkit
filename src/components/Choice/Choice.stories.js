'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.CustomControlled = exports.NativeFormControlled = exports.Controlled = exports.TallContent = exports.Basic = void 0;
var tslib_1 = require('tslib');
var addon_knobs_1 = require('@storybook/addon-knobs');
var react_1 = tslib_1.__importStar(require('react'));
var action_1 = require('../../storybook-helpers/action/action');
var Choice_1 = require('../Choice');
var README_md_1 = tslib_1.__importDefault(require('./README.md'));
exports.default = {
  title: 'Components/Choice',
  component: Choice_1.Choice,
  parameters: { readme: README_md_1.default },
};
var Basic = function() {
  return react_1.default.createElement(Choice_1.Choice, {
    onChange: (0, action_1.action)('onChange'),
    label: (0, addon_knobs_1.text)('label', 'Choice'),
    radio: (0, addon_knobs_1.boolean)('radio', false),
    isSwitch: (0, addon_knobs_1.boolean)('isSwitch', false),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    leftLabel: (0, addon_knobs_1.boolean)('leftLabel', false),
    center: (0, addon_knobs_1.boolean)('center', false),
    disabledLabelChange: (0, addon_knobs_1.boolean)(
      'disabledLabelChange',
      false,
    ),
  });
};
exports.Basic = Basic;
var TallContent = function() {
  return react_1.default.createElement(Choice_1.Choice, {
    onChange: (0, action_1.action)('onChange'),
    label: react_1.default.createElement(
      'div',
      null,
      react_1.default.createElement('div', null, 'Some'),
      react_1.default.createElement('div', null, 'Tall'),
      react_1.default.createElement('div', null, 'Content'),
    ),
    radio: (0, addon_knobs_1.boolean)('radio', false),
    isSwitch: (0, addon_knobs_1.boolean)('isSwitch', false),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    leftLabel: (0, addon_knobs_1.boolean)('leftLabel', false),
    center: (0, addon_knobs_1.boolean)('center', false),
    disabledLabelChange: (0, addon_knobs_1.boolean)(
      'disabledLabelChange',
      false,
    ),
  });
};
exports.TallContent = TallContent;
var Controlled = function() {
  return react_1.default.createElement(Choice_1.Choice, {
    onChange: (0, action_1.action)('onChange'),
    label: (0, addon_knobs_1.text)('label', 'Choice'),
    radio: (0, addon_knobs_1.boolean)('radio', false),
    isSwitch: (0, addon_knobs_1.boolean)('isSwitch', false),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    checked: (0, addon_knobs_1.boolean)('checked', false),
    leftLabel: (0, addon_knobs_1.boolean)('leftLabel', false),
    center: (0, addon_knobs_1.boolean)('center', false),
    disabledLabelChange: (0, addon_knobs_1.boolean)(
      'disabledLabelChange',
      false,
    ),
  });
};
exports.Controlled = Controlled;
var NativeFormControlled = function() {
  var props = {
    radio: (0, addon_knobs_1.boolean)('radio', true),
    isSwitch: (0, addon_knobs_1.boolean)('isSwitch', false),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    leftLabel: (0, addon_knobs_1.boolean)('leftLabel', false),
    center: (0, addon_knobs_1.boolean)('center', false),
    disabledLabelChange: (0, addon_knobs_1.boolean)(
      'disabledLabelChange',
      false,
    ),
  };
  return react_1.default.createElement(
    'form',
    { className: 'ui__base' },
    react_1.default.createElement(
      'p',
      null,
      'One from each group can be selected (native form behavior)',
    ),
    react_1.default.createElement('h3', null, 'Pet'),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, { label: 'Dog', value: 'dog', name: 'pet' }),
    ),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, { label: 'Cat', value: 'cat', name: 'pet' }),
    ),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, { label: 'Rat', value: 'rat', name: 'pet' }),
    ),
    react_1.default.createElement('h3', null, 'Color'),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Red',
        value: 'red',
        name: 'color',
      }),
    ),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Blue',
        value: 'blue',
        name: 'color',
      }),
    ),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Green',
        value: 'green',
        name: 'color',
      }),
    ),
  );
};
exports.NativeFormControlled = NativeFormControlled;
var CustomControlled = function() {
  var _a = (0, react_1.useState)(0),
    selected = _a[0],
    setSelected = _a[1];
  var props = {
    radio: (0, addon_knobs_1.boolean)('radio', true),
    isSwitch: (0, addon_knobs_1.boolean)('isSwitch', false),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    leftLabel: (0, addon_knobs_1.boolean)('leftLabel', false),
    center: (0, addon_knobs_1.boolean)('center', false),
    disabledLabelChange: (0, addon_knobs_1.boolean)(
      'disabledLabelChange',
      false,
    ),
  };
  return react_1.default.createElement(
    'div',
    { className: 'ui__base' },
    react_1.default.createElement(
      'p',
      null,
      'React state allows for only one to be selected (custom)',
    ),
    react_1.default.createElement('h3', null, 'Pet'),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Dog',
        onChange: function() {
          return setSelected(0);
        },
        checked: selected === 0,
      }),
    ),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Cat',
        onChange: function() {
          return setSelected(1);
        },
        checked: selected === 1,
      }),
    ),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Rat',
        onChange: function() {
          return setSelected(2);
        },
        checked: selected === 2,
      }),
    ),
    react_1.default.createElement('h3', null, 'Color'),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Red',
        onChange: function() {
          return setSelected(3);
        },
        checked: selected === 3,
      }),
    ),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Blue',
        onChange: function() {
          return setSelected(4);
        },
        checked: selected === 4,
      }),
    ),
    react_1.default.createElement(
      Choice_1.Choice,
      tslib_1.__assign({}, props, {
        label: 'Green',
        onChange: function() {
          return setSelected(5);
        },
        checked: selected === 5,
      }),
    ),
  );
};
exports.CustomControlled = CustomControlled;
//# sourceMappingURL=Choice.stories.js.map
