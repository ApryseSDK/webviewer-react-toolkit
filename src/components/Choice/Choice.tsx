import classnames from 'classnames';
import React, {
  ChangeEvent,
  forwardRef,
  InputHTMLAttributes,
  ReactNode,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from 'react';
import { useAccessibleFocus, useFocus } from '../../hooks';
import { useID } from '../../hooks/useID';
import { Remove } from '../../utils';
import { Icon } from '../Icon';

export interface ChoiceProps extends Remove<InputHTMLAttributes<HTMLInputElement>, 'type'> {
  /**
   * Label is optional, but it is encouraged to add the `aria-label` prop if you
   * are not labeling the choice, or if label is not a string.
   */
  label?: ReactNode;
  /**
   * If true, label will appear to the left of the choice.
   */
  leftLabel?: boolean;
  /**
   * Choice is a checkbox unless radio is true, in which case it is a radio
   * button.
   */
  radio?: boolean;
  /**
   * Choice becomes a switch, rather than a checkbox or radio. It will still
   * behave as either a checkbox, or radio if `radio` is true.
   */
  isSwitch?: boolean;
  /**
   * If true, choice will be centered vertically in the content.
   */
  center?: boolean;
}

export const Choice = forwardRef<HTMLInputElement, ChoiceProps>(
  (
    { label, leftLabel, className, children, id, radio, isSwitch, center, onChange, onFocus, onBlur, ...props },
    ref,
  ) => {
    const inputRef = useRef<HTMLInputElement>(null);
    useImperativeHandle(ref, () => inputRef.current as HTMLInputElement);

    const isUserTabbing = useAccessibleFocus();
    const { focused, handleOnFocus, handleOnBlur } = useFocus(onFocus, onBlur);

    const choiceID = useID(id);

    const [checked, setChecked] = useState<boolean>(() => props.checked ?? inputRef.current?.checked ?? false);
    useEffect(() => {
      if (props.checked !== undefined) setChecked(props.checked);
    }, [props.checked]);

    const handleOnChange = (event: ChangeEvent<HTMLInputElement>) => {
      if (props.checked === undefined) setChecked(event.target.checked);
      onChange?.(event);
    };

    // HACK: since there is no way to detect that a radio button is being
    // unchecked due to form activity, we subscribe all radio buttons to an
    // observable. When checked changes and props.checked is undefined, this
    // will trigger the observable which will ping all other subscribers to
    // check if they have become unchecked, and if so change their internal
    // state.
    useEffect(() => {
      if (props.name && radio) {
        return observable.subscribe(props.name, () => {
          if (inputRef.current && inputRef.current.checked !== checked) {
            setChecked(inputRef.current.checked);
          }
        });
      }
      return;
    }, [checked, props.name, radio]);

    const choiceClass = classnames(
      'ui__base ui__choice',
      {
        'ui__choice--radio': radio,
        'ui__choice--leftLabel': leftLabel,
        'ui__choice--checked': checked,
        'ui__choice--disabled': props.disabled,
        'ui__choice--center': center,
      },
      className,
    );

    const inputClass = classnames('ui__choice__input', { 'ui__choice__input--switch': isSwitch });

    const checkClass = isSwitch
      ? classnames('ui__choice__input__switch', {
          'ui__choice__input__switch--checked': checked,
          'ui__choice__input__switch--focus': isUserTabbing && focused,
        })
      : classnames('ui__choice__input__check', {
          'ui__choice__input__check--checked': checked,
          'ui__choice__input__check--focus': isUserTabbing && focused,
        });

    const labelElement = useMemo(() => {
      if (!label) return undefined;
      return (
        <label className="ui__choice__label" htmlFor={choiceID}>
          {label}
        </label>
      );
    }, [choiceID, label]);

    return (
      <span className={choiceClass}>
        {leftLabel ? labelElement : undefined}
        <span className={inputClass}>
          {isSwitch ? (
            <div className={checkClass}>
              <div className="ui__choice__input__toggle" />
            </div>
          ) : (
            <div className={checkClass}>
              {checked && !radio ? <Icon icon="Check" className="ui__choice__input__icon" /> : undefined}
            </div>
          )}
          <input
            {...props}
            id={choiceID}
            type={radio ? 'radio' : 'checkbox'}
            onChange={handleOnChange}
            ref={inputRef}
            onFocus={handleOnFocus}
            onBlur={handleOnBlur}
          >
            {children}
          </input>
        </span>
        {!leftLabel ? labelElement : undefined}
      </span>
    );
  },
);

/**
 * Observable helper to notify radio buttons that they have become unchecked.
 */
class RadioObservable {
  private _subscribers: { name: string; subscriber: () => void }[];

  constructor() {
    this._subscribers = [];
  }

  subscribe(name: string, subscriber: () => void) {
    this._trigger(name);

    this._subscribers.push({ name, subscriber });
    return this._unsubscribe(subscriber);
  }

  private _trigger(name: string) {
    this._subscribers.forEach((s) => {
      if (name === s.name) s.subscriber();
    });
  }

  private _unsubscribe(subscriber: () => void) {
    return () => {
      this._subscribers = this._subscribers.filter((s) => s.subscriber !== subscriber);
    };
  }
}

const observable = new RadioObservable();
