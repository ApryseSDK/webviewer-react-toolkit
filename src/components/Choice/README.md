Choice provides a component for either checkboxes or radio selection buttons.
These function how native radio or checkbox components would.
