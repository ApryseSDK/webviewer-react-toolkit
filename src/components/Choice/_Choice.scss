$choice-check-size: 15px;

$choice-switch-size: 14px;
$choice-switch-height: 10px;
$choice-switch-width: 24px;

.ui__choice {
  display: flex;
  margin: $padding-tiny $padding-tiny $padding-tiny 0;

  &--center {
    align-items: center;
  }

  &--disabled {
    opacity: 0.7;

    .ui__choice__label,
    .ui__choice__input > input {
      cursor: default;
    }
  }

  &--radio {
    .ui__choice__input__check {
      border-radius: 50%;

      &--checked {
        background-color: $color-contrast-1;
        border: 5px solid $color-blue-gray-5;
      }
    }
  }

  &__input {
    position: relative;
    height: $choice-check-size;
    width: $choice-check-size;
    flex-shrink: 0;

    &__check {
      position: absolute;
      width: $choice-check-size;
      height: $choice-check-size;
      padding: 2px;
      border: 1px solid $color-blue-gray-3;
      border-radius: $border-radius-small;
      background-color: $color-gray-1;
      transition: $focus-transition;

      &--checked {
        background-color: $color-blue-gray-5;
        border-color: $color-blue-gray-5;
      }

      &--focus {
        @include focus();
      }
    }

    &__switch {
      position: absolute;
      width: $choice-switch-width;
      height: $choice-switch-height;
      padding: 2px;
      border: 1px solid $color-blue-gray-3;
      border-radius: $choice-switch-height / 2;
      background-color: $color-gray-1;
      transition: $focus-transition;
      top: 50%;
      transform: translateY(-50%);

      &--checked .ui__choice__input__toggle {
        left: $choice-switch-width - $choice-switch-size - 1px;
      }

      &--focus .ui__choice__input__toggle {
        @include focus();
      }
    }

    &__toggle {
      position: absolute;
      height: $choice-switch-size;
      width: $choice-switch-size;
      border-radius: 50%;
      background-color: $color-blue-gray-5;
      top: 50%;
      left: -1px;
      transform: translateY(-50%);
      transition: left 0.25s ease;
    }

    &__icon {
      height: 17px;
      width: 17px;
      fill: $color-contrast-1;
      position: absolute;
      top: -3px;
      left: -2px;
    }

    > input {
      cursor: pointer;
      opacity: 0;
      position: absolute;
      height: $choice-check-size;
      width: $choice-check-size;
    }

    &--switch {
      height: $choice-check-size; // Keep as full-height to center switch.
      width: $choice-switch-width;

      > input {
        height: $choice-switch-size;
        width: $choice-switch-width;
      }
    }
  }

  &--leftLabel {
    .ui__choice__label {
      padding-right: $padding-tiny;
      padding-left: 0;
    }
  }

  &__label {
    cursor: pointer;
    padding-left: $padding-tiny;
    user-select: none;
  }
}
