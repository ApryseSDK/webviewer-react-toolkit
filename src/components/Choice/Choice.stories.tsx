import { boolean, text } from '@storybook/addon-knobs';
import React, { useState } from 'react';
import { action } from '../../storybook-helpers/action/action';
import { Choice } from '../Choice';
import readme from './README.md';

export default { title: 'Components/Choice', component: Choice, parameters: { readme } };

export const Basic = () => (
  <Choice
    onChange={action('onChange')}
    label={text('label', 'Choice')}
    radio={boolean('radio', false)}
    disabled={boolean('disabled', false)}
  />
);

export const Controlled = () => (
  <Choice
    onChange={action('onChange')}
    label={text('label', 'Choice')}
    radio={boolean('radio', false)}
    disabled={boolean('disabled', false)}
    checked={boolean('checked', false)}
  />
);

export const NativeFormControlled = () => {
  const props = {
    radio: boolean('radio', true),
    disabled: boolean('disabled', false),
  };

  return (
    <form className="ui__base">
      <p>One from each group can be selected (native form behavior)</p>
      <h3>Pet</h3>
      <Choice {...props} label="Dog" value="dog" name="pet" />
      <Choice {...props} label="Cat" value="cat" name="pet" />
      <Choice {...props} label="Rat" value="rat" name="pet" />

      <h3>Color</h3>
      <Choice {...props} label="Red" value="red" name="color" />
      <Choice {...props} label="Blue" value="blue" name="color" />
      <Choice {...props} label="Green" value="green" name="color" />
    </form>
  );
};

export const CustomControlled = () => {
  const [selected, setSelected] = useState(0);

  const props = {
    radio: boolean('radio', true),
    disabled: boolean('disabled', false),
  };

  return (
    <>
      <p>React state allows for only one to be selected (custom)</p>
      <h3>Pet</h3>
      <Choice {...props} label="Dog" onChange={() => setSelected(0)} checked={selected === 0} />
      <Choice {...props} label="Cat" onChange={() => setSelected(1)} checked={selected === 1} />
      <Choice {...props} label="Rat" onChange={() => setSelected(2)} checked={selected === 2} />

      <h3>Color</h3>
      <Choice {...props} label="Red" onChange={() => setSelected(3)} checked={selected === 3} />
      <Choice {...props} label="Blue" onChange={() => setSelected(4)} checked={selected === 4} />
      <Choice {...props} label="Green" onChange={() => setSelected(5)} checked={selected === 5} />
    </>
  );
};
