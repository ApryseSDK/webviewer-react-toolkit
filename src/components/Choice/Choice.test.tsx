import { shallow } from 'enzyme';
import React from 'react';
import { Choice } from '../Choice';

describe('Choice component', () => {
  it('renders its contents', () => {
    const choice = shallow(<Choice label="Label" />);
    expect(choice.find('.ui__choice')).toHaveLength(1);
  });

  it('snapshot renders default choice', () => {
    const choice = shallow(<Choice label="Label" />);
    expect(choice).toMatchSnapshot();
  });
});
