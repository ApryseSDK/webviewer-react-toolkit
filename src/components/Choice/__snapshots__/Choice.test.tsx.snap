// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choice component snapshot renders default choice 1`] = `
<span
  className="ui__base ui__choice"
>
  <span
    className="ui__choice__input"
  >
    <div
      className="ui__choice__input__check"
    />
    <input
      id="label_1"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="checkbox"
    />
  </span>
  <label
    className="ui__choice__label"
    htmlFor="label_1"
  >
    Label
  </label>
</span>
`;
